### ISSUE TITLE
Feature Proposal: add an ability to don't throw ConditionTimeoutException

### ISSUE THREAD

**Issue by Spikhalskiy** [jayway/awaitility#48]
Hi,

Sometimes we need to wait for something, but on the same time, we don't know if condition should be met.
Example:
We have an application that writes logs. This logs flushed asynchronously. In our test we want to check a state of the application log, which could be empty and it could be fine. So, we need to wait for... 5 seconds, for example, stop wait if log became non-empty or just continue after 2 seconds if the log is still empty.

Now we need to deal with ConditionTimeoutException which will be thrown if the condition didn't meet after awaiting period, which is ugly and makes code messy.

Proposal:
Add .silent() method to awaitility building chain which disables throwing the exception.

Optional:
Add result value to .until() method to make it possible to understand if the condition has been met during waiting without one more explicit check

What do you think?

**Comment by johanhaleby** [comment]
Hi,

Doesn't this feature already exists if you use "ignoreException", https://github.com/jayway/awaitility/wiki/Usage#example-9---ignoring exceptions?

**Comment by Spikhalskiy** [comment]
@johanhaleby No, by tutorial "you can choose to ignore exceptions while evaluating a condition". So, by adding ignoreException you ignore exceptions during condition checking, but don't ignore awaitility own exceptions like ConditionTimeoutException.

**Comment by chedlin** [comment]
I like this.  What if we could provide an onFailure(callable) that would let use do whatever we please?  In my case it would be fail("Some useful message"), so onFailure(() -> fail("Message")), or you could do onFailure(() -> return;), etc?  If not specified, it keeps the exception?

**Comment by johanhaleby** [comment]
Interesting, I like this better than "silent". Wouldn't a variant of Function (the java 8 functional interface) be sufficient for onTimeout? One could then decide to return "null", throw AssertionError or return an arbitrary value.

**Comment by johanhaleby** [comment]
For example there would be no need to do onTimeout(() -> fail("..")), just do onTimeout(ctx -> throw new AssertionError(..)) which will be default behavior. "ctx" contains data that one would need to construct a good error message or to create another type of return value.

**Comment by Spikhalskiy** [comment]
Would be good to have
*) still an easy way to ignore. It would be not very cool to write something like this ```onTimeout(() -> return;)``` each time. So, generic ```onTimeout()``` + alias for do nothing is the best option for me.
*) would be great to have a boolean return value if we don't throw an exception, so we can use awaitility statement directly in a ```while()``` for example

**Comment by chedlin** [comment]
I will admit I deal with a lot of legacy code.  The variant of Function sounds like a good way to go.  ctx would be good.

**Comment by mxk1235** [comment]
this feature would actually be pretty handy for us.  our use case is a function that we retry until it stops throwing exceptions.  if await() has exhausted the poll duration, we would like to see the exceptions that happened during retries, or *at the very least* have ConditionTimeoutException have a meaningful getCause().  For example, whether timeout last saw a SQLConnectionException or a different failing assert.

**Comment by srdo** [comment]
I think it would be helpful to have a way to specify that Awaitility should throw AssertionError instead of ConditionTimeoutException. It would make using Awaitility much less awkward in cases where we want the test to fail if the timeout is not met.

The `onTimeout` hook idea is great. Could consider adding a couple of convenience methods as well, e.g. `.silent` and `onTimeoutFail`.

The sample tests at https://github.com/awaitility/awaitility/blob/master/awaitility-java8-test/src/test/java/org/awaitility/AwaitilityJava8Test.java are a good example why this would be helpful. Most of them set the timeout through JUnit rather than Awaitility, so they don't have to try-catch-convert-to-AssertionError ConditionTimeoutException. This has the same issue as using e.g. `@Test(expected=...)` would, namely that you can't set the timeout locally to one condition, you have to set it for the entire test, and if the test fails you don't get useful feedback on the failed condition (e.g. "condition value was x, expected y").

**Comment by desember** [comment]
Hi guys,
Is there any chance that this feature will be implemented? It would be really nice to be able to specify your own callback for timeout with, for example, sensible output to log.

**Comment by johanhaleby** [comment]
Looks like it's requested by several people so I suppose it's something
that we should look into.

**Comment by ngbalk** [comment]
I've found Awaitility in the context of JUnit tests to be the most salient use-case, so I think natively supporting configuration to throw AssertionError instead of ConditionTimeoutException would be incredibly useful.  Supporting `.onTimeout` might be a generic solution, but I suspect an explicit configuration (i.e. `.throwAssertionErrorOnFailure` or similar) would be welcome :)

**Comment by olijaun** [comment]
Hi

Hmmm... the above mentioned commit is from 2018... Any chance to get this into master soon? Would be great!

Regards
Oliver

**Comment by vvscher** [comment]
I'm just working on a case where I need to poll for some time and NOT fail at the end.
I'd like to have a custom onTimeout(Supplier) handler  :)

**Comment by iklimchuk** [comment]
2 years later...

**Comment by Spikhalskiy** [comment]
@iklimchuk I heard they accept contributions

**Comment by Shmayro** [comment]
2023 ...

**Comment by zhaojc** [comment]
2024 still use try catch

**Comment by jonatan-ivanov** [comment]
I think Micrometer could also use this feature. Our use-case is coming from the behavior of Netty: based on protocol and OS, certain things can happen or not happen asynchronously in our tests. Instead of coming up the whole matrix of cases and handle them conditionally, we could say that let's try to wait for Netty to do its thing but in case it does not, that's ok, we can ignore the `ConditionTimeoutException`.

**Comment by jonatan-ivanov** [comment]
I created https://github.com/awaitility/awaitility/pull/299 to resolve this.
